contracts/core/OriginCore.sol

```diff
@@ -144,43 +121,109 @@ contract OriginCore is OriginCoreInterface, OriginCoreConfig {
+     *                          on the auxiliary chain.
+     * @param _auxiliaryBlockHash The block hash where the meta-block closes
+     *                          on the auxiliary chain.
      * @param _gas The total consumed gas on auxiliary within this meta-block.
+     * @param _originDynasty Dynasty of origin block within latest meta-block
+     *                          reported at auxiliary chain.
+     * @param _originBlockHash Block hash of origin block within latest
+     *                          meta-block reported at auxiliary chain.
      * @param _transactionRoot The transaction root of the meta-block. A trie
      *                         created by the auxiliary block store from the
      *                         transaction roots of all blocks.
-     * @param _auxiliaryDynasty The dynasty number where the meta-block closes
-     *                          on the auxiliary chain.
-     *
      * @return `true` if the proposal succeeds.
      */
     function proposeBlock(
         uint256 _height,
-        bytes32 _parent,
-        address[] _updatedValidators,
-        uint256[] _updatedWeights,
-        bytes20 _coreIdentifier,
+        bytes32 _coreIdentifier,
+        bytes32 _kernelHash,
+        uint256 _auxiliaryDynasty,
         bytes32 _auxiliaryBlockHash,
         uint256 _gas,
-        bytes32 _transactionRoot,
-        uint256 _auxiliaryDynasty
+        uint256 _originDynasty,
+        bytes32 _originBlockHash,
+        bytes32 _transactionRoot
     )
         external
-        returns (bool success_)
+        returns (bool)
     {
-        revert("Method not implemented.");
+
+        require(
+            _kernelHash != bytes32(0),
+            'Kernel hash should not be `0`.'
+        );
+
+        require(
+            _originDynasty > 0,
+            'Origin dynasty should not be `0`.'
+        );
+
+        require(
+            _originBlockHash != bytes32(0),
+            'Origin block should not be `0`.'
+        );
+
+        require(
+            _transactionRoot != bytes32(0),
+            'Transaction Root hash should not be `0`.'
+        );
+
+        require(
+            _coreIdentifier == auxiliaryCoreIdentifier,
+            'CoreIdentifier should be same as auxiliary core identifier.'
+        );
+
+        /* header of last meta block */
+        MetaBlock.Header memory latestMetaBlockHeader = reportedHeaders[head];
+
+        require(
+            latestMetaBlockHeader.kernel.height + 1 == _height,
+            'Height should be one more than last meta-block.'
+        );
+
+        require(
+            _auxiliaryDynasty > latestMetaBlockHeader.transition.auxiliaryDynasty,
+            'Auxiliary dynasty should be greater than last meta-block auxiliary dynasty.'
+        );
+
+        require(
+            _gas > latestMetaBlockHeader.transition.gas,
+            'Gas consumed should be greater than last meta-block gas.'
```

Comment:
Maybe we should call it "accumulated gas" in this contract to make it clear that it is increasing with every meta-block.